### PR TITLE
[Model Runner V2] Support logit_bias, allowed_token_ids, min_tokens

### PR DESCRIPTION
**Pull request by WoosukKwon**
> [!NOTE]
> Implements GPU-side sampling constraints with a dedicated state and Triton kernel.
> 
> - Adds `LogitBiasState` to stage per-request `allowed_token_ids`, `logit_bias`, and `min_tokens`/`stop_token_ids` using `UvaBackedTensor` and `StagedWriteTensor`
> - Introduces `_bias_kernel` (Triton) to: mask logits to only `allowed_token_ids`, apply per-token `logit_bias`, and enforce `min_tokens` by suppressing `stop_token_ids` until `min_len`
> - Provides `add_request`, `apply_staged_writes`, and `apply_logit_bias` to integrate with batched decoding; enforces max sizes via `MAX_NUM_*` constants
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2265d7200835fc0c64a1aa71d0cbf8c16a249d4b. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Implements GPU-side sampling constraints and enables staged float writes.
> 
> - Adds `LogitBiasState` managing per-request `allowed_token_ids`, `logit_bias` (ids + float biases), and `min_tokens`/`stop_token_ids` via `UvaBackedTensor`/`StagedWriteTensor`
> - New Triton `_bias_kernel` masks logits to `allowed_token_ids`, applies per-token `logit_bias`, and suppresses `stop_token_ids` until `min_len`
> - Exposes `add_request`, `apply_staged_writes`, and `apply_logit_bias` for batched decoding; enforces caps via `MAX_NUM_*`
> - Extends `StagedWriteTensor` to accept `torch.float32` and staged `list[float]` contents
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 3ea29ce7464b85e27b82a49801c62d95520ab592. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces GPU-side sampling constraints and staged float writes.
> 
> - Adds `LogitBiasState` managing per-request `allowed_token_ids`, `logit_bias` (ids + float biases), and `min_tokens`/`stop_token_ids` via `UvaBackedTensor`/`StagedWriteTensor`
> - New Triton `_bias_kernel` masks logits to only `allowed_token_ids`, applies per-token `logit_bias`, and suppresses `stop_token_ids` until `min_len`
> - Provides `add_request`, `apply_staged_writes`, `apply_logit_bias` for batched decoding; caps sizes via `MAX_NUM_*`
> - Extends `StagedWriteTensor` to support `torch.float32` and staged `list[float]` contents
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 28c59970519e6a1c5f771ac7505f1defc04f943d. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Implements GPU-side sampling constraints and staged float writes for batched decoding.
> 
> - Adds `LogitBiasState` to manage per-request `allowed_token_ids`, `logit_bias` (token ids + float biases), and `min_tokens`/`stop_token_ids` using `UvaBackedTensor`/`StagedWriteTensor`
> - Introduces Triton `_bias_kernel` to mask logits to `allowed_token_ids`, apply per-token bias, and suppress `stop_token_ids` until `min_len`; used by `apply_logit_bias`
> - Provides `add_request` and `apply_staged_writes` to stage/copy state; enforces caps via `MAX_NUM_*`
> - Extends `StagedWriteTensor` to support `torch.float32` and staged `list[float]` contents
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 970a74d23cdefe774b7744b2975b8ffc2083da70. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces GPU-side sampling constraints with a dedicated state and Triton kernel.
> 
> - Adds `LogitBiasState` managing per-request `allowed_token_ids`, `logit_bias` (ids + float biases), and `min_tokens`/`stop_token_ids` via `UvaBackedTensor`/`StagedWriteTensor`; exposes `add_request`, `apply_staged_writes`, `apply_logit_bias`
> - Implements `_bias_kernel` to mask logits to `allowed_token_ids`, apply per-token `logit_bias`, and suppress `stop_token_ids` until `min_len`
> - Extends `StagedWriteTensor` to accept `torch.float32` and stage `Iterable[float]` contents
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit c925021add945d31c73b3fa74d11acabdf047e34. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: vllm/v1/worker/gpu/sample/logit_bias.py [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        stop_token_ids = sampling_params.stop_token_ids
+        if stop_token_ids:
+            num_stop_token_ids = len(stop_token_ids)
+            if num_stop_token_ids > MAX_NUM_STOP_TOKEN_IDS:
+                raise ValueError(
+                    f"Too many stop tokens: {num_stop_token_ids}. "
+                    f"The max size is {MAX_NUM_STOP_TOKEN_IDS}."
+                )
+            self.num_stop_token_ids.np[req_idx] = num_stop_token_ids
+            self.stop_token_ids.apply_write(req_idx, 0, stop_token_ids)
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/logit_bias.py]: Good catch.

[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/logit_bias.py]: Good catch.

File: vllm/v1/worker/gpu/sample/logit_bias.py [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+            self.logit_bias_token_ids.stage_write(req_idx, 0, list(logit_bias.keys()))
+            self.logit_bias.stage_write(req_idx, 0, list(logit_bias.values()))
+        else:
+            self.num_logit_bias.np[req_idx] = 0
+
+        # Min tokens.
+        min_tokens = sampling_params.min_tokens
+        min_len = prompt_len + min_tokens
+        self.min_lens.np[req_idx] = min_len
+        stop_token_ids = sampling_params.stop_token_ids
```
[the remark this answers is not in the document]
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/logit_bias.py]: Nice catch. Fixed.

File: vllm/v1/worker/gpu/buffer_utils.py [hunk 3]
```diff
@@ -128,7 +129,7 @@ def __init__(
             self.num_rows, dtype=torch.int32, max_concurrency=max_concurrency
         )
 
-    def stage_write(self, index: int, start: int, x: list[int]) -> None:
+    def stage_write(self, index: int, start: int, x: list[int] | list[float]) -> None:
```
Comment (njhill) [inline vllm/v1/worker/gpu/buffer_utils.py]:
```suggestion
    def stage_write(self, index: int, start: int, x: Iterable[int] | Iterable[float]) -> None:
```

File: vllm/v1/worker/gpu/sample/logit_bias.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        if logit_bias:
+            num_logit_bias = len(logit_bias)
+            if num_logit_bias > MAX_NUM_LOGIT_BIAS_TOKENS:
+                raise ValueError(
+                    f"Too many logit bias tokens: {num_logit_bias}. "
+                    f"The max size is {MAX_NUM_LOGIT_BIAS_TOKENS}."
+                )
+            self.num_logit_bias.np[req_idx] = num_logit_bias
+            self.logit_bias_token_ids.stage_write(req_idx, 0, list(logit_bias.keys()))
+            self.logit_bias.stage_write(req_idx, 0, list(logit_bias.values()))
```
Comment (njhill) [inline vllm/v1/worker/gpu/sample/logit_bias.py]: no need to copy to list
```suggestion
            self.logit_bias_token_ids.stage_write(req_idx, 0, logit_bias.keys())
            self.logit_bias.stage_write(req_idx, 0, logit_bias.values())
```

**Review by njhill — COMMENTED** [review]
nice work!

File: vllm/v1/worker/gpu/buffer_utils.py (hunk 3 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/buffer_utils.py]: great idea. Fixed.

File: vllm/v1/worker/gpu/sample/logit_bias.py (hunk 4 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu/sample/logit_bias.py]: Fixed!
